This script aims to add Stadtteil information

In [1]:
import geopandas as gpd
from shapely.geometry import Point
import pandas as pd
import os

In [2]:
# Define input, output paths
current_dir = globals().get('current_dir', os.getcwd())
input_path = os.path.join(current_dir, '../data/mobidata/raw/csv_files/')
map_path = os.path.join(current_dir, '../data/map/Stadtteile_SHP/Stadtviertel_Karlsruhe.shp')
output_path = os.path.join(current_dir, '../data/mobidata/processed/')

# Define file path
input_file = 'dynamische_tafeln.csv'
input_file_path = os.path.join(input_path, input_file)

In [3]:
# 1) Load points (your mobility data)
df = pd.read_csv(input_file_path)
gdf_points = gpd.GeoDataFrame(
    df,
    geometry=gpd.points_from_xy(df["geometry_x"], df["geometry_y"]),
    crs="EPSG:25832",   # your points are in EPSG:25832
)

In [15]:
df_list = df.to_dict('records')
df_list

[{'type': 'FeatureCollection',
  'totalFeatures': 3,
  'numberMatched': 3,
  'numberReturned': 3,
  'timeStamp': '2025-10-05T21:25:23.058Z',
  'crs_type': 'name',
  'crs_properties_name': 'urn:ogc:def:crs:EPSG::25832',
  'feature_type': 'Feature',
  'feature_id': 'dynamische_tafeln.1',
  'geometry_name': 'geom',
  'geometry_type': 'Point',
  'geometry_coordinates': '[365587.89796653, 5366344.03078287]',
  'geometry_x': 365587.89796653,
  'geometry_y': 5366344.03078287,
  'id': 1,
  'standort': 'Pont des Bas',
  'text_fr': '  ',
  'text_de': '  ',
  'schilder_id': 1,
  'stand': '2025-01-08T08:39:12Z'},
 {'type': 'FeatureCollection',
  'totalFeatures': 3,
  'numberMatched': 3,
  'numberReturned': 3,
  'timeStamp': '2025-10-05T21:25:23.058Z',
  'crs_type': 'name',
  'crs_properties_name': 'urn:ogc:def:crs:EPSG::25832',
  'feature_type': 'Feature',
  'feature_id': 'dynamische_tafeln.2',
  'geometry_name': 'geom',
  'geometry_type': 'Point',
  'geometry_coordinates': '[378624.82493555, 5400

In [5]:
# 2) Load Stadtteile polygons
gdf_stadtteile = gpd.read_file(map_path)

In [6]:
gdf_stadtteile

OBJECTID NUMMER                        NAME  BEVD    SHAPE_AREA  \
0         47    132                      Bulach   0.0  2.349710e+06   
1         48    122                    Waldlage   0.0  6.152226e+05   
2         49    115  Neue Heidenstückersiedlung   0.0  7.117038e+05   
3         50    112             Hardecksiedlung   0.0  4.739133e+05   
4         51    111              Alt-Grünwinkel   0.0  1.141327e+06   
..       ...    ...                         ...   ...           ...   
65        66    071             Nördlicher Teil   0.0  1.377227e+06   
66        67    195                         Aue   0.0  2.187323e+06   
67        68    191                 Alt-Durlach   0.0  5.583243e+06   
68        69    161                    Waldlage   0.0  9.477397e+06   
69        70    171                Alt-Rintheim   0.0  3.123360e+06   

       SHAPE_LEN                                           geometry  
0    8835.931194  POLYGON ((454615.549 5426795.403, 454620.722 5...  
1    3868.385783  POLYGON ((453802.311 5426791.226, 453799.052 5...  
2    4693.388222  POLYGON ((453219.232 5427203.653, 453219.566 5...  
3    3617.545619  POLYGON ((453910.998 5427616.596, 453907.044 5...  
4    4960.769418  POLYGON ((453643.475 5428138.041, 453640.67 54...  
..           ...                                                ...  
65   5216.090139  POLYGON ((458149.345 5430144.52, 458151.172 54...  
66   7553.954377  POLYGON ((460626.689 5425626.578, 460624.636 5...  
67  13576.175961  POLYGON ((462708.966 5428419.393, 462706.385 5...  
68  14302.383936  POLYGON ((460151.186 5432841.279, 460150.673 5...  
69   9674.491623  POLYGON ((458850.82 5430323.717, 458854.788 54...  

[70 rows x 7 columns]

In [7]:

# Keep only the fields we need and give clear names
gdf_stadtteile = gdf_stadtteile.rename(columns={
    "NAME": "Stadtteil",
    "NUMMER": "Stadtteil_Nummer"
})

In [8]:
gdf_stadtteile

OBJECTID Stadtteil_Nummer                   Stadtteil  BEVD    SHAPE_AREA  \
0         47              132                      Bulach   0.0  2.349710e+06   
1         48              122                    Waldlage   0.0  6.152226e+05   
2         49              115  Neue Heidenstückersiedlung   0.0  7.117038e+05   
3         50              112             Hardecksiedlung   0.0  4.739133e+05   
4         51              111              Alt-Grünwinkel   0.0  1.141327e+06   
..       ...              ...                         ...   ...           ...   
65        66              071             Nördlicher Teil   0.0  1.377227e+06   
66        67              195                         Aue   0.0  2.187323e+06   
67        68              191                 Alt-Durlach   0.0  5.583243e+06   
68        69              161                    Waldlage   0.0  9.477397e+06   
69        70              171                Alt-Rintheim   0.0  3.123360e+06   

       SHAPE_LEN                                           geometry  
0    8835.931194  POLYGON ((454615.549 5426795.403, 454620.722 5...  
1    3868.385783  POLYGON ((453802.311 5426791.226, 453799.052 5...  
2    4693.388222  POLYGON ((453219.232 5427203.653, 453219.566 5...  
3    3617.545619  POLYGON ((453910.998 5427616.596, 453907.044 5...  
4    4960.769418  POLYGON ((453643.475 5428138.041, 453640.67 54...  
..           ...                                                ...  
65   5216.090139  POLYGON ((458149.345 5430144.52, 458151.172 54...  
66   7553.954377  POLYGON ((460626.689 5425626.578, 460624.636 5...  
67  13576.175961  POLYGON ((462708.966 5428419.393, 462706.385 5...  
68  14302.383936  POLYGON ((460151.186 5432841.279, 460150.673 5...  
69   9674.491623  POLYGON ((458850.82 5430323.717, 458854.788 54...  

[70 rows x 7 columns]

In [9]:
gdf_stadtteile.crs

<Projected CRS: EPSG:25832>
Name: ETRS89 / UTM zone 32N
Axis Info [cartesian]:
- E[east]: Easting (metre)
- N[north]: Northing (metre)
Area of Use:
- name: Europe between 6°E and 12°E: Austria; Denmark - onshore and offshore; Germany - onshore and offshore; Italy - onshore and offshore; Norway including Svalbard - onshore and offshore; Spain - offshore.
- bounds: (6.0, 36.53, 12.01, 84.01)
Coordinate Operation:
- name: UTM zone 32N
- method: Transverse Mercator
Datum: European Terrestrial Reference System 1989 ensemble
- Ellipsoid: GRS 1980
- Prime Meridian: Greenwich

In [10]:
# 3) Ensure both layers share the same CRS
if gdf_stadtteile.crs is None:
    # Only set if you are certain the polygons are in EPSG:25832
    gdf_stadtteile = gdf_stadtteile.set_crs("EPSG:25832")
elif gdf_stadtteile.crs.to_epsg() != 25832:
    gdf_stadtteile = gdf_stadtteile.to_crs(25832)

In [11]:
gdf_stadtteile

OBJECTID Stadtteil_Nummer                   Stadtteil  BEVD    SHAPE_AREA  \
0         47              132                      Bulach   0.0  2.349710e+06   
1         48              122                    Waldlage   0.0  6.152226e+05   
2         49              115  Neue Heidenstückersiedlung   0.0  7.117038e+05   
3         50              112             Hardecksiedlung   0.0  4.739133e+05   
4         51              111              Alt-Grünwinkel   0.0  1.141327e+06   
..       ...              ...                         ...   ...           ...   
65        66              071             Nördlicher Teil   0.0  1.377227e+06   
66        67              195                         Aue   0.0  2.187323e+06   
67        68              191                 Alt-Durlach   0.0  5.583243e+06   
68        69              161                    Waldlage   0.0  9.477397e+06   
69        70              171                Alt-Rintheim   0.0  3.123360e+06   

       SHAPE_LEN                                           geometry  
0    8835.931194  POLYGON ((454615.549 5426795.403, 454620.722 5...  
1    3868.385783  POLYGON ((453802.311 5426791.226, 453799.052 5...  
2    4693.388222  POLYGON ((453219.232 5427203.653, 453219.566 5...  
3    3617.545619  POLYGON ((453910.998 5427616.596, 453907.044 5...  
4    4960.769418  POLYGON ((453643.475 5428138.041, 453640.67 54...  
..           ...                                                ...  
65   5216.090139  POLYGON ((458149.345 5430144.52, 458151.172 54...  
66   7553.954377  POLYGON ((460626.689 5425626.578, 460624.636 5...  
67  13576.175961  POLYGON ((462708.966 5428419.393, 462706.385 5...  
68  14302.383936  POLYGON ((460151.186 5432841.279, 460150.673 5...  
69   9674.491623  POLYGON ((458850.82 5430323.717, 458854.788 54...  

[70 rows x 7 columns]

In [12]:
# 4) Spatial join (points within polygons)
# Keep only the column we need from gdf_stadtteile
gdf_stadtteile_subset = gdf_stadtteile[["Stadtteil_Nummer","Stadtteil", "geometry"]]

# Spatial join
gdf_joined = gpd.sjoin(gdf_points, gdf_stadtteile_subset, how="left", predicate="within")

In [13]:
# 5) If some points lie exactly on borders and return NaN, try a tiny buffer:
if gdf_joined["Stadtteil"].isna().any():
    # buffer(0) cleans invalid geometries; buffer(0.1) grows by 0.1m in EPSG:25832 units
    gdf_points_buf = gdf_points.copy()
    gdf_points_buf["geometry"] = gdf_points_buf.buffer(0.1)
    gdf_joined_fallback = gpd.sjoin(gdf_points_buf, gdf_stadtteile, how="left", predicate="intersects")
    # fill only the missing ones
    for col in ["Stadtteil", "Stadtteil_Nummer"]:
        gdf_joined[col] = gdf_joined[col].fillna(gdf_joined_fallback[col])



/var/folders/v0/n055j8sd1d1dv_bpn3tp_1n40000gn/T/ipykernel_18333/1816743394.py:9: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  gdf_joined[col] = gdf_joined[col].fillna(gdf_joined_fallback[col])


In [14]:
# 6) Display the joined variable
gdf_joined

type  totalFeatures  numberMatched  numberReturned  \
0  FeatureCollection              3              3               3   
1  FeatureCollection              3              3               3   
2  FeatureCollection              3              3               3   

                  timeStamp crs_type          crs_properties_name  \
0  2025-10-05T21:25:23.058Z     name  urn:ogc:def:crs:EPSG::25832   
1  2025-10-05T21:25:23.058Z     name  urn:ogc:def:crs:EPSG::25832   
2  2025-10-05T21:25:23.058Z     name  urn:ogc:def:crs:EPSG::25832   

  feature_type           feature_id geometry_name  ... id      standort  \
0      Feature  dynamische_tafeln.1          geom  ...  1  Pont des Bas   
1      Feature  dynamische_tafeln.2          geom  ...  2       Saverne   
2      Feature  dynamische_tafeln.3          geom  ...  3     Schirmeck   

   text_fr  text_de  schilder_id                 stand  \
0                              1  2025-01-08T08:39:12Z   
1                              2  2025-01-08T08:33:44Z   
2                              3  2025-01-08T07:39:47Z   

                         geometry index_right  Stadtteil_Nummer Stadtteil  
0  POINT (365587.898 5366344.031)         NaN               NaN       NaN  
1  POINT (378624.825 5400873.882)         NaN               NaN       NaN  
2  POINT (367399.304 5371448.902)         NaN               NaN       NaN  

[3 rows x 24 columns]

🤔 Need to be further checked, all Stadtteil_Nummer values are NaN...

### Logic of `sjoin(..., predicate="within")` 

You call:

```python
gdf_joined = gpd.sjoin(gdf_points, gdf_stadtteile_subset, how="left", predicate="within")
```

where:

* `gdf_points` = your **points** (each row is one mobility location).
* `gdf_stadtteile_subset` = your **polygons** (each row is one Stadtteil; keep only `Stadtteil_Nummer` and `geometry`).
* `how="left"` = keep **all** points (even if they don’t match any polygon → they get NaN).
* `predicate="within"` = a point matches a polygon **only if it lies strictly inside** (not on the boundary).

### The 3 steps GeoPandas performs

1. **Spatial index & candidate filter (fast)**

   * GeoPandas builds a spatial index (STRtree) on the polygons.
   * For each point, it finds polygon **candidates** whose bounding boxes overlap the point (super fast).

2. **Exact geometric test (precise)**

   * For each point–candidate pair, it checks the **“within”** relation using GEOS (DE-9IM).
   * “within” means the point is inside the polygon interior (if a point lies exactly on an edge/vertex, it’s **not** “within”).

3. **Tabular merge**

   * For every point that passed the test with some polygon, it **adds** that polygon’s attributes to the point’s row.
   * If **no** polygon matches → the polygon columns are `NaN`.
   * If (rare) a point matches **multiple** polygons (overlapping polygons), you’ll get **duplicate point rows**, one per polygon.

GeoPandas adds an `index_right` column showing the index of the matching polygon. You can drop it.

### Why CRS matters here

Both layers must be in the **same CRS** (you’re using EPSG:25832). If not, the coordinates won’t line up and no matches will be found (or you’ll get nonsense).

### Boundary cases (important!)

* Because `predicate="within"` is **strict**, points **on the boundary** return `NaN`.
* If you want boundary-inclusive behavior, use:

  * `predicate="intersects"` (widest net), or
  * `predicate="within"` first, then fill misses with a second join using `predicate="intersects"`, or
  * `predicate="covers"` (if your GeoPandas/pyogrio stack supports it).

### Applying this to your concrete data

* Your `gdf_points` rows look like:

  * e.g., point at `(365587.89796653, 5366344.03078287)` in EPSG:25832.
* Your `gdf_stadtteile` has polygons like “Alt-Grünwinkel”, “Alt-Rintheim”, etc., with column `NUMMER` (e.g., `171`) and `NAME`.

Do this cleanly:

```python
# Keep just what you need from polygons, with a friendly name
gdf_stadtteile_subset = gdf_stadtteile.rename(
    columns={"NUMMER": "Stadtteil_Nummer"}
)[["Stadtteil_Nummer", "geometry"]]

# Spatial join: label each point with its Stadtteil_Nummer
gdf_joined = gpd.sjoin(
    gdf_points,
    gdf_stadtteile_subset,
    how="left",
    predicate="within"   # strict interior; boundary points may be NaN
)

# Optional cleanup
gdf_joined = gdf_joined.drop(columns=["index_right"], errors="ignore")
```

### Handling tricky points (optional)

If you suspect some points lie exactly on borders (or there’s floating-point noise), try a tiny buffer on points and a boundary-inclusive predicate:

```python
# Tiny buffer in meters (since EPSG:25832 units are meters)
gdf_points_buf = gdf_points.copy()
gdf_points_buf["geometry"] = gdf_points_buf.buffer(0.05)  # 5 cm

gdf_joined_b = gpd.sjoin(
    gdf_points_buf, gdf_stadtteile_subset, how="left", predicate="intersects"
).drop(columns=["index_right"], errors="ignore")

# Fill only missing Stadtteil_Nummer from the buffered join
gdf_joined["Stadtteil_Nummer"] = gdf_joined["Stadtteil_Nummer"].fillna(
    gdf_joined_b["Stadtteil_Nummer"]
)
```

### Quick sanity checks

* Confirm CRS on both layers is EPSG:25832.
* Use `gdf_joined["Stadtteil_Nummer"].isna().mean()` to see % unmatched.
* Plot a few points over the polygon map to visually confirm correctness.

That’s the full story of how `sjoin(..., predicate="within")` evaluates each of your mobility points and tags them with the correct *Stadtteil*.
